lib/util/Ranges.v3

```diff
@@ -78,38 +78,38 @@ component Ranges {
 	}
 	// Performs binary search and returns the index.
 	def binarySearch<T>(range: Range<T>, val: T, lt: (T, T) -> bool) -> int {
-    var l = 0, r = range.length;
-    while (l < r) {
-      var mid = (l + r) >> 1;
-      var mid_val = range[mid];
-      if (lt(val, mid_val)) r = mid - 1;
-      else if (lt(mid_val, val)) l = mid + 1;
-      else return mid;
-    }
-    return -1;
-  }
-  // Returns the *first* element that is *not less* than `val` in a sorted range.
-  def lowerbound<T>(range: Range<T>, val: T, lt: (T, T) -> bool) -> int {
-    var start = 0;
-    var count = range.length;
-    while (count > 0) {
-      var step = count >> 1;
-      var mid = start + step;
-      if (lt(range[mid], val)) { start = mid + 1; count -= step + 1; }
-      else count = step;
-    }
-    return start;
-  } 
-  // Returns the *first* element that is *greater* than `val` in a sorted range.
-  def upperbound<T>(range: Range<T>, val: T, lt: (T, T) -> bool) -> int {
-    var start = 0;
-    var count = range.length;
-    while (count > 0) {
-      var step = count >> 1;
-      var mid = start + step;
-      if (!lt(val, range[mid])) { start = mid + 1; count -= step + 1; }
-      else count = step;
-    }
-    return start;
-  }
+		var l = 0, r = range.length;
+		while (l < r) {
+			var mid = (l + r) >> 1;
+			var mid_val = range[mid];
+			if (lt(val, mid_val)) r = mid - 1;
+			else if (lt(mid_val, val)) l = mid + 1;
+			else return mid;
+		}
+		return -1;
+	}
+	// Returns the *first* element that is *not less* than `val` in a sorted range.
+	def lowerbound<T>(range: Range<T>, val: T, lt: (T, T) -> bool) -> int {
+		var start = 0;
+		var count = range.length;
+		while (count > 0) {
+			var step = count >> 1;
+			var mid = start + step;
+			if (lt(range[mid], val)) { start = mid + 1; count -= step + 1; }
+			else count = step;
+		}
+		return start;
+	} 
+	// Returns the *first* element that is *greater* than `val` in a sorted range.
+	def upperbound<T>(range: Range<T>, val: T, lt: (T, T) -> bool) -> int {
+		var start = 0;
+		var count = range.length;
+		while (count > 0) {
+			var step = count >> 1;
+			var mid = start + step;
+			if (!lt(val, range[mid])) { start = mid + 1; count -= step + 1; }
```

Comment:
I think there is a subtle distinction between `!lt(x, y)` and `lt(y, x)`. I think this should code should use the latter. I think it should be accompanied with a comment that illustrates, like so:

```
// Assumes that {range} is sorted according to `lt` and returns the first element that is *greater* than `val`.
// e in         | elem  elem  elem  elem  elem  elem  |
// lt(e, val)   | true  true  false false false false |
// lt(val, e)   | false false false true  true  true  |
//                                  ^
```

I think all three methods could use this comment.